### PR TITLE
typescript visualization ports

### PR DESCRIPTION
**Pull request by pascal4mile**
WIP

### CODE REVIEW

**Comment by pascal4mile** [comment]
@statico Can you give this a review and let me know if you see any glaring issues?

File: TODO.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+put everything on EC2
+code cleanup
+steps to replicate config race condtion
+steps to replicate infinite query loop
+create PR to our master
+describe typescript d3.selection quandary
```
Comment (statico) [inline TODO.md]: You probably don't want to leave this checked in.

File: package.json [hunk 2]
```diff
@@ -27,11 +27,13 @@
...
-    "yarn": "1.3.2"
+    "yarn": "^1.3.2"
   },
   "dependencies": {
     "@types/d3": "^5.0.0",
     "@types/d3-sankey": "^0.7.2",
-    "d3-sankey": "^0.7.1"
+    "d3-sankey": "^0.7.1",
+    "global": "^4.3.2",
+    "watch": "^1.0.2"
```
Comment (statico) [inline package.json]: I don't think `watch` needs to be added to the dependencies. A `watch` script that uses webpack's watch feature would be lighter weight:

`    "watch": "webpack --watch --progress --config webpack.config.js",`

File: src/examples/chord/chord.ts [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+// Global values provided via the API
+declare var looker: Looker
+declare var LookerCharts: LookerChartUtils
+
+interface ChordVisualization extends VisualizationDefinition {
+  svg?: any,
+  tooltip?: any,
+  computeMatrix: (data: VisData, dimensions: string[], measure: string) => any,
+  titleText: (lookup: any, source: any, target: any, formatter: any) => string,
+}
```
Comment (statico) [inline src/examples/chord/chord.ts]: Can we do any better than the `any` types here? Just for the sake of TypeScript correctness.

File: src/examples/chord/chord.ts [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    const width = element.clientWidth
+    const height = element.clientHeight
+    const margin = 10
+    const thickness = 15
+    const outerRadius = Math.min(width, height) * 0.5
+    const innerRadius = outerRadius - thickness
+
+    // Stop if radius is < 0
+    // TODO: show warning to user ???
+    // TODO: Set a min-radius ???
```
Comment (statico) [inline src/examples/chord/chord.ts:132]: Possibly remove TODOs?

File: src/examples/chord/chord.ts [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    if (innerRadius < 0) return
+
+    const valueFormatter = formatType(measure.value_format) || ((s: any): string => s.toString())
+
+    const tooltip = this.tooltip
+
+    // Set color scale
+    const color = d3.scaleOrdinal().range(config.color_range)
+    // const color = d3.scaleOrdinal().range(config.color_range || this.options.color_range.default) // DNR
```
Comment (statico) [inline src/examples/chord/chord.ts]: Remove commented out code?

File: src/examples/common/utils.ts [hunk 6]
```diff
@@ -6,6 +6,10 @@ import {
   VisualizationDefinition
 } from '../types/types'
 
+export function log(...args: any[]) {
+  console.log.apply(console, args)
```
Comment (statico) [inline src/examples/common/utils.ts:10]: FWIW, you can also pass splats/spreads in: `console.log(...args)`

File: src/examples/liquid_fill_gauge/README.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+#  Liquid Fill Gauge
+
+![](liquid_fill_gauge.png)
+
+This diagram displays a liquid fill gauge (LFG), displaying either a single measure value as a percentage, or a comparison of one measure to another measure.
+
+
+****
+add .mov file once we have the lfg working on the looker hosted instance
```
Comment (statico) [inline src/examples/liquid_fill_gauge/README.md]: Remove TODO?

File: src/examples/liquid_fill_gauge/liquid_fill_gauge.ts [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+declare var require: any
+declare var _: any
+
+import * as d3 from 'd3'
+import { handleErrors } from '../common/utils'
+/**
+ * TODO install this version?
+ * https://github.com/ugomeda/d3-liquid-fill-gauge
+ * hmm, it's not published on npm
+ */
```
Comment (statico) [inline src/examples/liquid_fill_gauge/liquid_fill_gauge.ts:11]: TODO here.

You can still add the repo as a dependency in package.json: https://stackoverflow.com/a/38731705

Alternatively this seems to be present in https://photino.github.io/d3-components/api/liquid-fill-gauge.html

File: src/examples/treemap/treemap.ts [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    const width = element.clientWidth
+    const height = element.clientHeight
+
+    const dimensions = queryResponse.fields.dimension_like
+    const measure = queryResponse.fields.measure_like[0]
+
+    const format = formatType(measure.value_format) || ((s: any): string => s.toString())
+
+    const color = d3.scaleOrdinal().range(config.color_range)
+    // const color = d3.scaleOrdinal().range(config.color_range || this.options.color_range.default) // DNR
```
Comment (statico) [inline src/examples/treemap/treemap.ts]: Commented code

File: src/examples/treemap/treemap.ts [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+      .attr('x', 4)
+
+    const root = d3.hierarchy(burrow(data)).sum((d: any) => {
+      return 'data' in d ? d.data[measure.name].value : 0
+    })
+    treemap(root)
+
+    const cell = svg.selectAll('.node')
+        .data(root.descendants())
+      .enter().append('g')
```
Comment (statico) [inline src/examples/treemap/treemap.ts]: Needs an indent

Comment (statico) [reply src/examples/treemap/treemap.ts]: Rather, a lot of this file is indented inconsistently compared to the rest of the project. Looks like some lines use tabs?

**Review by statico — APPROVED** [review]
Great stuff. I've left some comments with varying levels of pedantry and usefulness — feel free to accept or ignore them.

File: src/examples/common/utils.ts (hunk 6 above)
Comment (pascal4mile) [reply src/examples/common/utils.ts:10]: yeah this is just my `log` shortcut so i don't have to type `console.log` : )
i'll remove this before we send it up to looker.

File: src/examples/treemap/treemap.ts (hunk 10 above)
Comment (pascal4mile) [reply src/examples/treemap/treemap.ts]: yeah, i think this was pre-existing in the JS version. It's following some d3 indentation conventions where certain chained methods (.enter, .exit) are outdented relative to the methods that follow them. I'm not really a fan of these conventions, myself, but I don't know if it's our place to make stylistic recommendations. Maybe we should get some guidance on this from looker. @nickfogler what do you think?

File: src/examples/treemap/treemap.ts [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+				.attr('y', (d) => {
+          return d.depth === 1 ? '13' : '10'
+        })
+				.attr('x', 2)
+				.style('font-family', 'Helvetica, Arial, sans-serif')
+        .style('fill', 'white')
+				.style('font-size', (d) => {
+          return d.depth === 1 ? '14px' : '10px'
+        })
+				.text((d) => d.data.name === 'root' ? '' : d.data.name)
```
Comment (pascal4mile) [inline src/examples/treemap/treemap.ts]: okay yeah this indentation is wack, not sure what happened here. will review.

Comment (pascal4mile) [reply src/examples/treemap/treemap.ts]: yeah those are tabs copied over from the older example. will fix.

File: src/examples/treemap/treemap.ts (hunk 10 above)
Comment (statico) [reply src/examples/treemap/treemap.ts]: Pretty sure their style is a 2-space indent, and there's a `tslint` config in this repo that they defined. I'd go with whatever makes `tslint` happy.

File: package.json (hunk 2 above)
Comment (pascal4mile) [reply package.json]: haha this is me doing `yarn add global watch` instead of `yarn global add watch`. will fix

**Comment by pascal4mile** [comment]
noticed there's a `lint-fix` script, so i ran that and address a few issues it raised. these are outstanding:

```
$ tslint --fix --project tsconfig.json  --format stylish 'src/**/*.ts'
/Users/pascal/Projects/Looker/custom_visualizations_v2/src/examples/chord/chord.ts
ERROR: 170:9   no-use-before-declare          variable 'ribbons' used before declaration
ERROR: 181:34  no-inferred-empty-object-type  Explicit type parameter needs to be provided to the function call
ERROR: 213:34  no-inferred-empty-object-type  Explicit type parameter needs to be provided to the function call

/Users/pascal/Projects/Looker/custom_visualizations_v2/src/examples/sunburst/sunburst.ts
ERROR: 93:11   no-unused-variable             'x' is declared but its value is never read.
ERROR: 94:11   no-unused-variable             'y' is declared but its value is never read.
ERROR: 140:14  no-inferred-empty-object-type  Explicit type parameter needs to be provided to the function call

/Users/pascal/Projects/Looker/custom_visualizations_v2/src/examples/treemap/treemap.ts
ERROR: 163:32  no-inferred-empty-object-type  Explicit type parameter needs to be provided to the function call
```

File: src/examples/chord/chord.ts (hunk 4 above)
Comment (pascal4mile) [reply src/examples/chord/chord.ts:132]: @nickfogler these TODOs were pre-existing in the JS version. should i keep them in or remove them?

File: src/examples/liquid_fill_gauge/README.md (hunk 7 above)
Comment (pascal4mile) [reply src/examples/liquid_fill_gauge/README.md]: @4mile-spencertaylor is this your TODO?

**Pull request merged**